### PR TITLE
test(public-profile): daily canary + admin ops surface (JOV-1872)

### PR DESCRIPTION
**Pull request by itstimwhite**
## Summary

- Playwright E2E spec `canary-public-profile.spec.ts` exercising `/tim`, `/tim/alerts`, `/tim/pay`, audience-visit endpoint, notification subscription form, claim view (`?noredirect=1`), and React hydration mismatch detector
- Daily cron `public-profile-canary` (06:13 UTC) hitting production `/tim` + audience-visit; emits Sentry breadcrumb + writes Redis key for the ops panel
- Admin ops panel status block: one-line pass/fail + last run timestamp (subtraction principle — compact, no new surfaces)
- 23 unit tests for all canary assertion helpers
- CRON_REGISTRY.md and vercel.json updated

Tracks **JOV-1872**. Parent **JOV-1855** (Reliability: Every Bug Becomes a Detector).

## Bug → Detector mapping

| Past bug | This canary catches it via |
|---|---|
| JOV-2199 — `audience_members.latest_referrer_url` column missing → 500 | `audience-visit endpoint health` check |
| JOV-2202 — React hydration mismatch on profile page | `no React hydration warnings` detector |
| JOV-2050 / JOV-2018 — Profile ISR 404 | `profile renders 200 + h1` check |

## Cost Impact

- **Daily cron**: 4 HTTP GET/POST calls/day to `jov.ie` ≈ **$0/month**
- **Scaling factor**: O(1) — fixed 4 calls regardless of user count
- **Storage**: one Redis key (TTL 26h, ~1 KB) ≈ $0
- Approved by JOV-1872 acceptance criteria

## Test plan

- [x] `pnpm --filter @jovie/web run typecheck` — clean
- [x] `pnpm biome check apps/web` — clean
- [x] Vitest: `lib/canaries/public-profile.test.ts` — 23/23 pass
- [x] All pre-commit + pre-push hooks pass
- [ ] Playwright canary spec against dev:web:browse with `/tim` seeded
- [ ] Cron route returns 200 in dev with `Authorization: Bearer $CRON_SECRET` header
- [ ] Admin ops panel renders canary status block

## Summary by CodeRabbit

* **New Features**
  * Daily public-profile canary endpoint that runs parallel health checks, records pass/fail, persists last run, and emits monitoring breadcrumbs/errors.
  * Admin Ops dashboard shows public-profile canary status (pass/fail/unknown), failing check names, and run timestamp.

* **Tests**
  * Added end-to-end canary tests covering profile, alerts, audience, and payment flows, plus deterministic fixtures.
  * Added unit tests for canary helpers and report formatting/validation.

* **Documentation**
  * Added cron registry and deployment schedule entry for the new health check.

[Review Change Stack](https://app.coderabbit.ai/change-stack/JovieInc/Jovie/pull/8913?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by itstimwhite** [comment]
## Review Summary
**Verdict**: Changes Requested

The PR is well-structured and follows most Jovie conventions. The canary infrastructure is thorough, with solid unit test coverage and a clear Playwright E2E spec. However, there is a critical gap in the cron route that could allow a 200‑response server error page to go undetected.

---

### Critical Issues

#### Cron route does not validate response body for server errors on `/tim/alerts` and `/tim/pay`
**`apps/web/app/api/cron/public-profile-canary/route.ts`** – lines 80–87  
**`apps/web/lib/canaries/public-profile.ts`** – lines 140–155 (`checkHttpGet`)

The cron calls `checkHttpGet` for the `alerts-200` and `pay-200` checks **without** supplying an `expectBodyFn`.  
In the current code, `checkHttpGet` returns `ok: true` as soon as the HTTP status is in the 2xx–3xx range — it never reads or inspects the response body for a server error indicator (`hasServerError`).

If a regression causes one of these pages to return a 200 status code but include an error message (e.g. an “Application error” fallback, an “Internal Server Error” rendered in the page body, or a React error boundary), the canary would incorrectly report a pass.  
This directly contradicts the bug-detector mapping described in the PR (JOV‑2199, JOV‑2202) and reduces confidence in the production monitoring loop.

**Fix**  
- Modify the `alerts-200` and `pay-200` calls to pass a default body verifier, for example:  
  ```ts
  body => !hasServerError(body)
  ```
- Alternatively, refactor `checkHttpGet` to **always** look for server errors in the body when no custom `expectBodyFn` is provided. This makes the default behaviour safer.
- Update the relevant unit tests to verify that `checkHttpGet` rejects a 200 response containing a known server error string.

---

### Warnings

*(None)*

---

### Suggestions

#### 1. Consider adding a body check for the audience‑visit POST as well
**`apps/web/lib/canaries/public-profile.ts`** – lines 170–192 (`checkHttpPost`)

Currently `checkHttpPost` only checks the status code (and allows 4xx). If the endpoint returns 200 but the body contains an internal error (e.g. a malformed response due to a failed DB query), the canary would not flag it.  
Since the audience‑visit endpoint is the detector for JOV‑2199, reading the body and verifying it is valid JSON would add a extra safety net.  
*(Not critical because the 404/400 cases are already handled, and a genuinely broken endpoint tends to return a 5xx.)*

#### 2. Add a unit test for `getPublicProfileCanaryStatus` in `lib/admin/ops-queries.ts`
The function is simple, but adding a test that mocks Redis (returning both a string and a pre‑parsed object) would protect against regressions in the Redis client or JSON.parse.

#### 3. Normalize the “no data yet” UI when Redis is empty
**`apps/web/app/app/(shell)/admin/ops/page.tsx`** – line 169  
When `canaryPass` is `null`, the block shows “No data yet” but the status dot is `bg-tertiary-token` (grey).  
Consider making the text colour slightly dimmer or adding a small label like “Awaiting first run” to differentiate from an actual pass/fail state.

---

### What Looks Good

- **Clean separation of concerns** – The canary helpers are pure functions, shared between the cron route and the E2E spec, with no Playwright dependency.
- **Thorough unit tests** – 23 tests covering `isOkStatus`, `hasServerError`, `buildVisitPayload`, `buildReport`, `formatReportSummary`, and constant values. Edge cases (case‑insensitivity, null bodies, UTC formatting) are covered.
- **Graceful degradation** – Redis unavailability is logged as a warning, and the ops panel returns `null` without crashing the page.
- **Security** – The cron endpoint is protected by `verifyCronRequest` (presumably requiring a secret header). No hardcoded secrets in the diff.
- **Admin ops integration** – The one‑line status block follows the “subtraction principle” (compact, no new surface), uses icons instead of emoji, and respects existing layout conventions.
- **Playwright E2E spec** – Includes hydration error detection, graceful skip when the profile is not seeded, and proper timeout handling.
- **Documentation** – `CRON_REGISTRY.md` and `vercel.json` are updated, and the route file has a thorough doc comment explaining purpose, costs, and schedule.

---

*Address the missing body-check in the cron route before merging.*

**Review by itstimwhite — COMMENTED** [review]
Automated code review from Hermes PR Review Bot (deepseek/deepseek-v4-flash).

**Comment by itstimwhite** [comment]
Addressed the review blocker and QA finding.

Changes pushed:
- `checkHttpGet` now reads successful GET bodies by default and fails on known server-error indicators even when callers do not provide a custom body assertion. This covers `/tim/alerts` and `/tim/pay` in the cron path.
- Added a unit regression test for a 200 response containing `Application error` without a custom assertion.
- Stabilized the Playwright canary so the DB-backed audience-visit check skips when local `DATABASE_URL` is unavailable instead of timing out; increased that DB-backed check timeout to match the other public-profile checks.

Local evidence:
- `pnpm --filter @jovie/web exec vitest run lib/canaries/public-profile.test.ts --config=vitest.config.mts` — 24 passed
- `pnpm --filter @jovie/web exec biome check lib/canaries/public-profile.ts lib/canaries/public-profile.test.ts` — passed
- `pnpm --filter @jovie/web run typecheck -- --pretty false` — passed after refreshing locked deps on the rebased main base
- Pre-push hook — full Biome, proxy guard, Tailwind guard, typecheck, lint passed
- `pnpm --filter @jovie/web exec playwright test tests/e2e/canary-public-profile.spec.ts --project=chromium --reporter=line` — 7 passed, 2 skipped locally because DB/auth prerequisites are absent

File: apps/web/lib/admin/ops-queries.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import 'server-only';
+
+import { CANARY_REDIS_KEY } from '@/app/api/cron/public-profile-canary/route';
+import type { CanaryReport } from '@/lib/canaries/public-profile';
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/lib/admin/ops-queries.ts]: Addressed in 4ff9b3339: CANARY_REDIS_KEY now lives in '@/lib/canaries/public-profile' and both the cron route and admin ops query import from that shared canary module.

File: apps/web/lib/canaries/public-profile.test.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    const result = await checkHttpGet(
+      'alerts-200',
+      'https://jov.ie/tim/alerts'
+    );
+
+    expect(result.ok).toBe(false);
+    expect(result.statusCode).toBe(200);
+    expect(result.detail).toBe('Response body contains server error indicator');
+  });
+});
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/lib/canaries/public-profile.test.ts:116]: Addressed in 4ff9b3339: added checkHttpPost unit coverage for 200, 429, 200-with-error-body, 404, 500 retry, and 307 redirect semantics. Focused Vitest passes: 30 tests.

File: apps/web/lib/canaries/public-profile.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  name: string,
+  url: string,
+  expectBodyFn?: (body: string) => boolean
+): Promise<CanaryCheckResult> {
+  const start = Date.now();
+  try {
+    const res = await fetch(url, {
+      signal: AbortSignal.timeout(10_000),
+      redirect: 'follow',
+    });
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/lib/canaries/public-profile.ts]: Addressed in 4ff9b3339: checkHttpGet and checkHttpPost now use fetchWithRetry with timeout, 3 attempts, and exponential backoff for transient network/5xx failures.

File: apps/web/lib/canaries/public-profile.ts [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    const res = await fetch(url, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(body),
+      signal: AbortSignal.timeout(10_000),
+    });
+    const durationMs = Date.now() - start;
+
+    // 429 is acceptable (rate-limited canary is a healthy system)
+    if (!isOkStatus(res.status) && res.status !== 429) {
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/lib/canaries/public-profile.ts]: Addressed in 4ff9b3339: checkHttpPost now accepts only 2xx or 429; 3xx redirects return ok=false with HTTP status detail.

File: apps/web/tests/e2e/canary-public-profile.spec.ts [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+  page: Page,
+  path: string
+): Promise<{ status: number; ok: boolean }> {
+  const response = await smokeNavigate(page, path, {
+    timeout: SMOKE_TIMEOUTS.NAVIGATION,
+  });
+  const status = response?.status() ?? 0;
+  if (status === 0 || status >= 500) {
+    return { status, ok: false };
+  }
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/tests/e2e/canary-public-profile.spec.ts]: Addressed in 4ff9b3339: goAndWait now treats 404 as ok=false so seeded-route skip logic is reachable.

File: docs/CRON_REGISTRY.md [hunk 6]
```diff
@@ -40,8 +40,9 @@ Source of truth: `vercel.json` (repo root — Vercel reads this file; `apps/web/
 | `/api/cron/process-pre-saves` | `0 2 * * *` | Daily at 02:00 UTC |
 | `/api/cron/monitor-metadata-submissions` | `0 * * * *` | Hourly |
 | `/api/cron/process-metadata-submissions` | `0 4 * * *` | Daily at 04:00 UTC |
+| `/api/cron/public-profile-canary` | `13 6 * * *` | Daily at 06:13 UTC |
 
-9 paths are currently scheduled in production. `cleanup-sms-intents` was folded into `daily-maintenance` as a sub-job per JOV-1901 (see AUTOMATION_AUDIT.md). Other cron route files exist as standalone endpoints whose logic is called as sub-jobs of `frequent` or `daily-maintenance`.
+10 paths are currently scheduled in production. `cleanup-sms-intents` was folded into `daily-maintenance` as a sub-job per JOV-1901 (see AUTOMATION_AUDIT.md). Other cron route files exist as standalone endpoints whose logic is called as sub-jobs of `frequent` or `daily-maintenance`.
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply docs/CRON_REGISTRY.md]: Addressed in 4ff9b3339: docs/CRON_REGISTRY.md now reports 11 scheduled production paths.

File: apps/web/lib/admin/ops-queries.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    const raw = await redis.get<string>(CANARY_REDIS_KEY);
+    if (!raw) return null;
+
+    // redis.get may return the value already parsed if the Upstash client
+    // detects JSON — handle both cases.
+    if (typeof raw === 'object') {
+      return raw as unknown as CanaryReport;
+    }
+
+    return JSON.parse(raw) as CanaryReport;
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/lib/admin/ops-queries.ts]: Addressed in 4ff9b3339: admin ops now parses Redis canary payloads as unknown and validates CanaryReport/check shapes before returning them.

File: apps/web/tests/e2e/canary-public-profile.spec.ts [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+
+    const { status, ok } = await goAndWait(page, CANARY_SPEC_ROUTES.profile);
+
+    if (!ok) {
+      hydrationErrors.cleanup();
+      if (status === 404) {
+        test.skip(true, '/tim not seeded');
+        return;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/tests/e2e/canary-public-profile.spec.ts:484]: Addressed in 4ff9b3339: the hydration test now explicitly fails non-404 load failures before checking hydration warnings.

**Comment by itstimwhite** [comment]
Review hardening slice pushed: 4ff9b3339

Addressed CodeRabbit comments:
- Moved CANARY_REDIS_KEY out of the API route import path and into shared canary helpers.
- Added fetchWithRetry for canary GET/POST with timeout, 3 attempts, and exponential backoff.
- Tightened checkHttpPost to accept only 2xx or 429, reject redirects, and reject 200 error bodies.
- Added POST helper unit coverage for 200, 429, error body, 404, 500 retry, and 307.
- Validated Redis canary payload shape before returning it to admin ops.
- Fixed goAndWait 404 handling and hydration-test non-404 failure assertion.
- Updated cron registry count to 11.

Local evidence:
- node --version: v22.22.1
- pnpm --version: 9.15.4
- pnpm biome check touched files: pass
- pnpm --filter @jovie/web exec vitest run lib/canaries/public-profile.test.ts --config=vitest.config.mts: 30 passed
- pnpm --filter @jovie/web run typecheck -- --pretty false: pass
- pnpm --filter @jovie/web exec playwright test tests/e2e/canary-public-profile.spec.ts --project=chromium --reporter=line: 7 passed, 2 skipped
- pre-push: full biome, proxy guard, Tailwind guard, typecheck, lint all passed

File: apps/web/tests/e2e/canary-public-profile.spec.ts [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+        timeout: 10_000,
+      })
+      .catch(() => null);
+
+    if (capturedStatus !== null) {
+      expect(
+        capturedStatus,
+        `Subscription API returned ${capturedStatus} — 5xx indicates a broken subscribe path`
+      ).toBeLessThan(500);
+    }
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/tests/e2e/canary-public-profile.spec.ts]: Fixed in d5f999cde. The subscription submit test now fails when no /api/audience response is captured before checking the response status. Verified with biome, typecheck, and the canary Playwright spec.

File: apps/web/tests/e2e/canary-public-profile.spec.ts [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+    if (!ok) {
+      if (status === 404) {
+        test.skip(true, '/tim profile not seeded in this environment');
+        return;
+      }
+      expect(
+        status,
+        `/${CANARY_CREATOR.handle} returned ${status}`
+      ).toBeLessThan(500);
+    }
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/tests/e2e/canary-public-profile.spec.ts:140]: Fixed in d5f999cde. Page navigation now treats any non-2xx/3xx status as not OK, keeps 404 as the only seeded-profile skip, and all non-404 page failures assert 200 <= status < 400 via expectPageSuccessStatus. Verified with biome, typecheck, and the canary Playwright spec.
